Allow TaskGroup's ChildTaskResult Type To Be Inferred (#2494)
swiftlang/swift-evolution: proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md

# Allow TaskGroup's ChildTaskResult Type To Be Inferred

* Proposal: [SE-NNNN](NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md)
* Author: [Richard L Zarth III](https://github.com/rlziii)
* Review Manager: TBD
* Status: **Awaiting review**

> * Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN)

Reviewer: ```suggestion
* Implementation: [apple/swift#74517](https://github.com/apple/swift/pull/74517)
```

* Review: ([pitch](https://forums.swift.org/t/allow-taskgroups-childtaskresult-type-to-be-inferred/72175))

## Introduction

> `TaskGroup` and `ThrowingTaskGroup` currently require that one of their two generics (`ChildTaskResult`) always be specified upon creation; this can be simplified by allowing the compiler to infer both of the generics in most cases.

Reviewer: ```suggestion
`TaskGroup` and `ThrowingTaskGroup` currently require that one of their two generics (`ChildTaskResult`) always be specified upon creation. Due to improvements in closure parameter/result type inference introduced by [SE-0326](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0326-extending-multi-statement-closure-inference.md) this can be simplified by allowing the compiler to infer both of the generics in most cases.
```

## Motivation

Currently to create a new task group, there are two generics involved: `ChildTaskResult` and `GroupResult`.  The latter can often be inferred in many cases, but the former must always be supplied as part of either the `withTaskGroup(of:returning:body:)` or `withThrowingTaskGroup(of:returning:body:)` function.  For example:

```swift
let s = withTaskGroup(of: Void.self) { group in
  group.addTask { /* ... */ }
  group.addTask { /* ... */ }
  group.addTask { /* ... */ }

  return "Hello, world!"
}
```

> The type of `s` (which is the `GroupResult` generic) is inferred above (as `String`).  However, the return value of the `addTask` closures cannot be inferred and must be supplied via `of childTaskResultType: ChildTaskResult.Type` (above as `Void`).

Reviewer: ```suggestion
The type of `s` (which is the `GroupResult` type) is correctly inferred as `String`.  However, the return value of the `addTask` closures cannot is not inferred and currently must be supplied to the `of:`parameter of the withTaskGroup function (e.g. `Void`).
```

It is also possible to create a compiler error by returning two different values from a `addTask` closure:

```swift
withTaskGroup { group in
    group.addTask { "Hello, world!" }
    group.addTask { 42 } // Cannot convert value of type 'Int' to closure result type 'String'
}
```

The compiler will already give a good error message here, since the first `addTask` statement is what determined (in this case) that the `ChildTaskResult` generic was set to `String`.  If this needs to be made more clear (instead of being inferred), the user can always specify the generic directly as before:

```swift
withTaskGroup(of: Int.self) { group in
    // Now the error has moved here since the generic was specified up front...
    group.addTask { "Hello, world!" } // Cannot convert value of type 'String' to closure result type 'Int'
    group.addTask { 42 }
}
```

## Source compatibility

Omitting the `of childTaskResultType: ChildTaskResult.Type` parameter for both `withTaskGroup(of:returning:body:)` and `withThrowingTaskGroup(of:returning:body:)` is new, and therefore the inference of `ChildTaskResult` is opt-in and does not break source compatibility.

## ABI compatibility

> No ABI impact since this is an additive change to default arguments.

Reviewer: ```suggestion
No ABI impact since adding a default argument value is binary compatible change.
```

## Implications on adoption

This feature can be freely adopted and un-adopted in source
code with no deployment constraints and without affecting source or ABI
compatibility.

## Future directions

> ...

Reviewer: Maybe let's add this:

---

### TaskGroup APIs without the "with..." closures

While not possible without more compiler features to enforce the safety of a task group not escaping a context, and having to await all of its results at the end of a scope... it is an interesting future direction to explore a TaskGroup API that does not need to resort to `with...` methods, like this:

```
// Potential long-term direction that might be possible
func test() async { 
  let group = TaskGroup<Int>()
  group.addTask { ... }
  
  // going out of scope would have to imply group.awaitAll()
}
```

If we were to explore such API, the type inference rules would be somewhat different, and a TaskGroup would likely be initialized more similarly to a collection: `TaskGroup<Int>`.

This proposal has no impact on this future direction, and can be accepted as is, without precluding future developments in API ergonomics like this.

## Alternatives considered

The main alternative is to do nothing; as in, leave the `withTaskGroup(of:returning:body:)` and `withThrowingTaskGroup(of:returning:body:)` APIs like they are and require the `ChildTaskResult` generic to always be specified.

## Acknowledgments

> ...

Reviewer: If you'd like to drop a small nod here for review/confirming viability to myself and @xedin that'd be nice, your call though :)

```swift
public func withTaskGroup<ChildTaskResult, GroupResult>(
    of childTaskResultType: ChildTaskResult.Type = ChildTaskResult.self, // <- Updated.
    returning returnType: GroupResult.Type = GroupResult.self,
    body: (inout TaskGroup<ChildTaskResult>) async -> GroupResult
) async -> GroupResult where ChildTaskResult : Sendable
```

This allows the original example above to be simplified:

```swift
let s = withTaskGroup { group in
  group.addTask { someVoidFunction() }
  group.addTask { someVoidFunction() }
  group.addTask { someVoidFunction() }

  return "Hello, world!"
}
```

In the above snippet, `ChildTaskResult` is inferred as `Void` and `GroupResult` is inferred as `String`.  Not needing to specify the generics explicitly will simplify the API design for these functions and make it easier for new users of these APIs, as it can currently be confusing to understand the differences between `ChildTaskResult` and `GroupResult` (especially when one or both of those is `Void`).

## Detailed design

> Because type inference is top-down, it relies on the first statement that uses `group` to infer determine the generic for `ChildTaskResult`.  Therefore, it is possible to get a compiler error by creating a task group where the first use of `group` does not use `addTask(...)`, like so:

Reviewer 2: Thank you! I'll kick the review off shortly.

```swift
// Expect `ChildTaskResult` to be `Void`...
withTaskGroup { group in // Generic parameter 'ChildTaskResult' could not be inferred
    // Since `addTask(...)` wasn't the first statement, this fails to compile.
    group.cancelAll()
    group.addTask { 42 }
}
```

This can be fixed by going back to specifying the generic like before:

```swift
// Expect `ChildTaskResult` to be `Int`...
withTaskGroup(of: Int.self) { group in
    group.cancelAll()

>     group.addTask { 42 }

Reviewer 2: I think it would be beneficial to have some more realistic examples. For example, there are small-but-realistic examples at https://www.avanderlee.com/concurrency/task-groups-in-swift/ and https://www.kodeco.com/books/modern-concurrency-in-swift/v1.0/chapters/7-concurrent-code-with-taskgroup that you could reference (with attribution!) and will give reviewers a better feel for how this impacts use of the feature.

Author: I have updated the examples to use more real-world scenarios.  Thank you for the suggestion.  👍
